### PR TITLE
new formula: icecream 1.1rc1

### PR DESCRIPTION
**Pull request by tduehr**
Adds icecream a distributed compiler toolchain.

### CODE REVIEW

File: Library/Formula/icecream.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
+class Icecream < Formula
+  homepage ""
+  url "https://github.com/icecc/icecream/archive/1.1rc1.tar.gz"
```
Comment (mistydemeo) [inline Library/Formula/icecream.rb]: Just to check, is this a stable release?

File: Library/Formula/icecream.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  version "1.1rc1"
```
Comment (mistydemeo) [inline Library/Formula/icecream.rb]: You can omit this, it'll be scanned from the URL.

File: Library/Formula/icecream.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
+class Icecream < Formula
+  homepage ""
```
Comment (mistydemeo) [inline Library/Formula/icecream.rb]: Can you add this, please?

File: Library/Formula/icecream.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  sha1 "ccde08f67297122710270b440492a528df042f52"
+
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
+  depends_on "docbook2X" => :build
+
+  def install
+    system "./autogen.sh"
+
```
Comment (mistydemeo) [inline Library/Formula/icecream.rb]: No newline here.

File: Library/Formula/icecream.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                          "--disable-silent-rules",
+                          "--prefix=#{prefix}"
+
+    system "make", "install"
+    (prefix+'org.opensuse.icecc.plist').write iceccd_plist
+    (prefix+'org.opensuse.icecc-scheduler.plist').write scheduler_plist
+  end
+
+  def caveats
+    return <<-EOS
```
Comment (mistydemeo) [inline Library/Formula/icecream.rb]: No need for an explicit `return`; in Ruby, the return value of a method/function is the last expression. Instead, just start this on the same line as `def`:

```ruby
def caveats; <<-EOS.undent
...
  EOS
end
```

Comment (mistydemeo) [inline Library/Formula/icecream.rb]: Incidentally, you can use the `undent` string method there to allow you to write the string indented at the same level as the rest of the code; the indentation will be removed for you.

File: Library/Formula/icecream.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+To override the toolset with icecc, add to your path:
+   #{prefix}/bin/icecc-toolset
```
Comment (mistydemeo) [inline Library/Formula/icecream.rb]: Does this install a directory into `bin`? `bin` should only contain executables (no directories or non-executable files).

File: Library/Formula/icecream.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    cp #{prefix}/org.opensuse.icecc-scheduler.plist ~/Library/LaunchAgents/
+    launchctl load -w ~/Library/LaunchAgents/org.opensuse.icecc-scheduler.plist
+
+If this is an upgrade, to restart a previous setup of the icecc daemon or scheduler:
+    launchctl unload -w ~/Library/LaunchAgents/org.openuses.icecc.plist
+    launchctl load -w ~/Library/LaunchAgents/org.openuses.icecc.plist
+ or
+    launchctl unload -w ~/Library/LaunchAgents/org.openuses.icecc-scheduler.plist
+    launchctl load -w ~/Library/LaunchAgents/org.openuses.icecc-scheduler.plist
+EOS
```
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: We don't need instructions for each of the plists; at least one of them will be printed out by `brew info` already.

File: Library/Formula/icecream.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  def install
+    system "./autogen.sh"
+
+    system "./configure", "--disable-debug",
+                          "--disable-dependency-tracking",
+                          "--disable-silent-rules",
+                          "--prefix=#{prefix}"
+
+    system "make", "install"
+    (prefix+'org.opensuse.icecc.plist').write iceccd_plist
```
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: Why `org.opensuse`?

Comment (tduehr) [inline Library/Formula/icecream.rb]: icecream came from opensuse. These plists were copied from an earlier PR to add icecream I just modified this part a little.

File: Library/Formula/icecream.rb (hunk 5 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: it seems I missed that return. I didn't realize undent was available here. It'll look much better that way.

File: Library/Formula/icecream.rb (hunk 6 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: I'll have to look into the best way to handle that. I'm not yet familiar with the tool chain but I expect that directory to have wrappers for things like `gcc` and `as`. Would `icecc-gcc` be acceptable?

File: Library/Formula/icecream.rb (hunk 7 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: Does that mean neither plist caveats are needed?

File: Library/Formula/icecream.rb (hunk 2 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: This was added by `brew create`. I'll remove.

File: Library/Formula/icecream.rb (hunk 1 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: it's a release candidate, but it compiles on osx without modification unlike 1.01.

File: Library/Formula/icecream.rb (hunk 6 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: it actually installs them into libexec. I'll change the caveat text.

File: Library/Formula/icecream.rb (hunk 7 above)
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: Test with `brew install` install and viewing the output.

File: Library/Formula/icecream.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: We'll probably want to wait for a stable version.

**Comment by tduehr** [comment]
I got 1.0.1 to install. I put 1.1rc1 in as a dev version. `brew` does not properly parse the version from the dev filename, so it is explicitly defined.

File: Library/Formula/icecream.rb [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  option "enable-clang-wrappers", "Use symlink wrappers for clang/clang++"
+  option "enable-clang-rewrite-includes", "Use by default Clang's -frewrite-includes option"
+
+  devel do
+    url "https://github.com/icecc/icecream/archive/1.1rc1.tar.gz"
+    version "1.1rc1"
+    sha1 "ccde08f67297122710270b440492a528df042f52"
+  end
+
+  stable do
```
Comment (DomT4) [inline Library/Formula/icecream.rb]: Stable should go above ` devel `.

File: Library/Formula/icecream.rb [hunk 10]
```diff
@@ -0,0 +1,97 @@
+class Icecream < Formula
+  homepage "https://en.opensuse.org/Icecream"
+
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
+  depends_on "docbook2X" => [:recommended, :build]
+  option "without-docbook2X", "Build without man page"
```
Comment (DomT4) [inline Library/Formula/icecream.rb]: ` options ` tend to need to go above dependencies.

**Comment by tduehr** [comment]
Okay, done

File: Library/Formula/icecream.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+  devel do
+    url "https://github.com/icecc/icecream/archive/1.1rc1.tar.gz"
+    version "1.1rc1"
+    sha1 "ccde08f67297122710270b440492a528df042f52"
+  end
+
+  fails_with :clang do
+    build 600
+    cause "requires gnu libstdc++"
+  end
```
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: Is there an upstream issue? I doubt we'll accept this as-is as it means it's basically not ported to OSX.

File: Library/Formula/icecream.rb [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        <true/>
+    </dict>
+    </plist>
+    EOS
+  end
+
+  test do
+    system "#{bin}/iceccd", "-h"
+    system "#{bin}/icecc", "--version"
+    system "#{bin}/icerun", "--version"
```
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Library/Formula/icecream.rb [hunk 13]
```diff
@@ -0,0 +1,97 @@
+class Icecream < Formula
+  homepage "https://en.opensuse.org/Icecream"
+
+  option "without-docbook2X", "Build without man page"
+  option "enable-clang-wrappers", "Use symlink wrappers for clang/clang++"
+  option "enable-clang-rewrite-includes", "Use by default Clang's -frewrite-includes option"
```
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: All options should start with `with` or `without`.

**Comment by mikemcquaid** [comment]
Lots of test failures: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/20997/testReport/

File: Library/Formula/icecream.rb (hunk 11 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: issue for libstdc++ or for what was preventing it from compiling under OSX?

Comment (DomT4) [inline Library/Formula/icecream.rb]: Wouldn't ` ENV.libstdcxx if ENV.compiler == :clang ` work instead of ` fails_with :clang `?

Comment (tduehr) [inline Library/Formula/icecream.rb]: that's what I couldn't find. trying now.

**Comment by tduehr** [comment]
The perl related failures are in docbook2X. Is there something I should be doing for those?

File: Library/Formula/icecream.rb (hunk 12 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: Yes, but not in a way that tests the distributed compilation.

File: Library/Formula/icecream.rb (hunk 11 above)
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: An upstream issue for what's preventing it compiling with `libc++`

File: Library/Formula/icecream.rb (hunk 12 above)
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: That's fine, it doesn't need to do that, just run a tool and make sure there isn't an immediate segfault doing something marginally more complex than `--help` or `--version

File: Library/Formula/icecream.rb [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
+  depends_on "docbook2X" => [:recommended, :build]
```
Comment (mikemcquaid) [inline Library/Formula/icecream.rb]: Should probably make this `:optional` given it requires manual intervention to install a perl lib.

File: Library/Formula/icecream.rb (hunk 11 above)
Comment (tduehr) [inline Library/Formula/icecream.rb]: `ENV.libstdcxx if ENV.compiler == :clang` does work and the problem has already been fixed upstream in 1.1rc1.

**Comment by tduehr** [comment]
passing now

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew and for being so tolerate with all the back-and-forth! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by tduehr** [comment]
NP, most of the back and forth was because I didn't read or forgot the formula rules...

**Pull request closed without merging**